view/db.go

```diff
@@ -0,0 +1,142 @@
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/squirrel"
+
+	"github.com/flanksource/duty/context"
+	"github.com/flanksource/duty/models"
+)
+
+// Row represents a single row of data mapped to view columns
+type Row []any
+
+func GetViewColumnDefs(ctx context.Context, namespace, name string) (ViewColumnDefList, error) {
+	var view models.View
+	err := ctx.DB().Where("namespace = ? AND name = ?", namespace, name).First(&view).Error
+	if err != nil {
+		return nil, err
+	}
+
+	var spec struct {
+		Columns []ViewColumnDef `json:"columns"`
+	}
+
+	err = json.Unmarshal(view.Spec, &spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return spec.Columns, nil
+}
+
+func GetAllViews(ctx context.Context) ([]models.View, error) {
+	var views []models.View
+	if err := ctx.DB().Where("deleted_at IS NULL").Find(&views).Error; err != nil {
+		return nil, err
+	}
+
+	return views, nil
+}
+
+func CreateViewTable(ctx context.Context, table string, columns []ViewColumnDef) error {
+	if ctx.DB().Migrator().HasTable(table) {
+		return nil
+	}
+
+	var columnDefs []string
+	for _, col := range columns {
+		colDef := fmt.Sprintf("%s %s", col.Name, getPostgresType(col.Type))
+		columnDefs = append(columnDefs, colDef)
+	}
+
+	columnDefs = append(columnDefs, "agent_id UUID NOT NULL DEFAULT '00000000-0000-0000-0000-000000000000'::uuid")
+	columnDefs = append(columnDefs, "is_pushed BOOLEAN DEFAULT FALSE")
+
+	sql := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (%s)", table, strings.Join(columnDefs, ", "))
+	return ctx.DB().Exec(sql).Error
+}
+
+func getPostgresType(colType ColumnType) string {
+	switch colType {
+	case ColumnTypeString:
+		return "TEXT"
+	case ColumnTypeNumber:
+		return "NUMERIC"
+	case ColumnTypeBoolean:
+		return "BOOLEAN"
+	case ColumnTypeDateTime:
+		return "TIMESTAMP WITH TIME ZONE"
+	case ColumnTypeDuration:
+		return "BIGINT"
+	case ColumnTypeHealth:
+		return "TEXT"
+	case ColumnTypeStatus:
+		return "TEXT"
+	case ColumnTypeGauge:
+		return "JSONB"
+	default:
+		return "TEXT"
+	}
+}
+
+func ReadViewTable(ctx context.Context, table string) ([]Row, error) {
+	rows, err := ctx.DB().Select("*").Table(table).Rows()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read view table (%s): %w", table, err)
+	}
+	defer rows.Close()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get columns for view table (%s): %w", table, err)
+	}
+
+	var viewRows []Row
+	for rows.Next() {
+		viewRow := make(Row, len(columns))
+		valuePtrs := make([]any, len(columns))
+		for i := range viewRow {
+			valuePtrs[i] = &viewRow[i]
+		}
+
+		if err := rows.Scan(valuePtrs...); err != nil {
+			return nil, fmt.Errorf("failed to scan row: %w", err)
+		}
+
+		viewRows = append(viewRows, viewRow)
+	}
+
+	return viewRows, nil
+}
+
+func InsertViewRows(ctx context.Context, table string, columns []ViewColumnDef, rows []Row) error {
+	if err := ctx.DB().Exec(fmt.Sprintf("TRUNCATE TABLE %s", table)).Error; err != nil {
```

Comment:
don't we need todo this in a transaction ?

Reply:
> TRUNCATE isn't MVCC safe, so some clients in other transactions will see the table as empty.

I don't think that's true. From what I've observed, other clients just wait and see the newly committed rows.

We can use `DELETE FROM` which doesn't block.